File: Frida/douyin_mystery_user.js
console.log("[*] 抖音神秘人主页地址获取插件已加载");

let mysteryUserCache = new Map();

function interceptRoomMessages() {
    try {
        const AwemeRoomMessageManager = ObjC.classes.AwemeRoomMessageManager;
        if (AwemeRoomMessageManager) {
            console.log("[+] 找到 AwemeRoomMessageManager 类");
            
            const handleMessage = AwemeRoomMessageManager['- handleMessage:'];
            if (handleMessage) {
                Interceptor.attach(handleMessage.implementation, {
                    onEnter: function(args) {
                        try {
                            const message = ObjC.Object(args[2]);
                            processMessage(message);
                        } catch (e) {
                            console.log("[-] 处理消息错误:", e);
                        }
                    }
                });
                console.log("[+] 已 hook handleMessage: 方法");
            }
        }
    } catch (e) {
        console.log("[-] hook 房间消息管理器失败:", e);
    }
}

function processMessage(message) {
    try {
        if (!message) return;
        
        const messageType = message.type ? message.type.toString() : "";
        
        if (messageType.includes("mystery") || messageType.includes("user")) {
            const userInfo = message.user;
            if (userInfo) {
                const userId = userInfo.uid ? userInfo.uid.toString() : "";
                const secUserId = userInfo.sec_uid ? userInfo.sec_uid.toString() : "";
                const nickname = userInfo.nickname ? userInfo.nickname.toString() : "神秘人";
                
                if (userId && secUserId && !mysteryUserCache.has(userId)) {
                    mysteryUserCache.set(userId, { secUid: secUserId, nickname: nickname });
                    
                    const homePageURL = `https://www.douyin.com/user/${secUserId}`;
                    
                    console.log("\n[!] ========== 发现神秘人 ==========");
                    console.log(`[!] 昵称: ${nickname}`);
                    console.log(`[!] UID: ${userId}`);
                    console.log(`[!] 主页地址: ${homePageURL}`);
                    console.log("[!] =================================\n");
                    
                    showNotification(nickname, homePageURL);
                }
            }
        }
    } catch (e) {
        console.log("[-] 解析消息错误:", e);
    }
}

function showNotification(title, url) {
    try {
        const UIAlertController = ObjC.classes.UIAlertController;
        const UIAlertAction = ObjC.classes.UIAlertAction;
        const UIApplication = ObjC.classes.UIApplication;
        
        const alert = UIAlertController.alertControllerWithTitle_message_preferredStyle_(
            `发现神秘人: ${title}`,
            url,
            0 
        );
        
        const copyAction = UIAlertAction.actionWithTitle_style_handler_(
            "复制链接",
            0,
            new ObjC.Block({
                retType: 'void',
                argTypes: ['object'],
                implementation: function() {
                    const UIPasteboard = ObjC.classes.UIPasteboard;
                    const pasteboard = UIPasteboard.generalPasteboard();
                    pasteboard.setString_(url);
                    console.log("[+] 链接已复制到剪贴板");
                }
            })
        );
        
        const okAction = UIAlertAction.actionWithTitle_style_handler_(
            "确定",
            1,
            new ObjC.Block({
                retType: 'void',
                argTypes: ['object'],
                implementation: function() {}
            })
        );
        
        alert.addAction_(copyAction);
        alert.addAction_(okAction);
        
        const app = UIApplication.sharedApplication();
        const window = app.keyWindow();
        const rootVC = window.rootViewController();
        
        if (rootVC) {
            rootVC.presentViewController_animated_completion_(alert, true, NULL);
        }
    } catch (e) {
        console.log("[-] 显示通知失败:", e);
    }
}

function interceptNetworkRequests() {
    try {
        const NSURLSession = ObjC.classes.NSURLSession;
        const NSURL = ObjC.classes.NSURL;
        
        Interceptor.attach(ObjC.classes.NSURLSession['- dataTaskWithRequest:completionHandler:'].implementation, {
            onEnter: function(args) {
                try {
                    const request = ObjC.Object(args[2]);
                    const url = request.URL().absoluteString().toString();
                    
                    if (url.includes("webcast") || url.includes("room") || url.includes("user")) {
                        console.log("[*] 请求:", url);
                    }
                } catch (e) {
                }
            }
        });
    } catch (e) {
    }
}

function exportMysteryUsers() {
    console.log("\n[*] ========== 已发现的神秘人列表 ==========");
    mysteryUserCache.forEach((info, uid) => {
        console.log(`[*] ${info.nickname} (UID: ${uid})`);
        console.log(`    主页: https://www.douyin.com/user/${info.secUid}`);
    });
    console.log("[*] ========================================\n");
}

if (ObjC.available) {
    setTimeout(() => {
        interceptRoomMessages();
        interceptNetworkRequests();
        console.log("[*] 插件初始化完成，等待神秘人出现...\n");
    }, 1000);
    
    rpc.exports = {
        getMysteryUsers: function() {
            let result = [];
            mysteryUserCache.forEach((info, uid) => {
                result.push({
                    uid: uid,
                    secUid: info.secUid,
                    nickname: info.nickname,
                    homePage: `https://www.douyin.com/user/${info.secUid}`
                });
            });
            return result;
        },
        exportList: function() {
            exportMysteryUsers();
            return "导出完成";
        }
    };
} else {
    console.log("[-] Objective-C runtime not available");
}

File: Dylib/DouyinMysteryUser.m
#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import <objc/runtime.h>
#import <dlfcn.h>

static NSMutableDictionary *mysteryUserInfoCache;
static IMP originalHandleMessage;

@interface MysteryUserHelper : NSObject
@end

@implementation MysteryUserHelper

+ (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *copyAction = [UIAlertAction actionWithTitle:@"复制链接" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            UIPasteboard.generalPasteboard.string = message;
        }];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:copyAction];
        [alert addAction:okAction];
        
        UIViewController *rootVC = [UIApplication sharedApplication].keyWindow.rootViewController;
        [rootVC presentViewController:alert animated:YES completion:nil];
    });
}

@end

static void hooked_handleMessage(id self, SEL _cmd, id message) {
    @try {
        if (message) {
            NSString *messageType = [NSString stringWithFormat:@"%@", [message valueForKey:@"type"]];
            if ([messageType containsString:@"mystery"] || [messageType containsString:@"user"]) {
                id userInfo = [message valueForKey:@"user"];
                if (userInfo) {
                    NSString *userId = [userInfo valueForKey:@"uid"];
                    NSString *secUserId = [userInfo valueForKey:@"sec_uid"];
                    NSString *nickname = [userInfo valueForKey:@"nickname"];
                    
                    if (userId && secUserId) {
                        if (!mysteryUserInfoCache) {
                            mysteryUserInfoCache = [NSMutableDictionary dictionary];
                        }
                        
                        if (!mysteryUserInfoCache[userId]) {
                            mysteryUserInfoCache[userId] = @{@"sec_uid": secUserId, @"nickname": nickname ?: @"神秘人"};
                            
                            NSString *homePageURL = [NSString stringWithFormat:@"https://www.douyin.com/user/%@", secUserId];
                            NSLog(@"[DouyinMysteryUser] 发现神秘人: %@ (UID: %@)", nickname ?: @"未知", userId);
                            NSLog(@"[DouyinMysteryUser] 主页地址: %@", homePageURL);
                            
                            [MysteryUserHelper showAlertWithTitle:@"发现神秘人" message:homePageURL];
                        }
                    }
                }
            }
        }
    } @catch (NSException *exception) {
        NSLog(@"[DouyinMysteryUser] 错误: %@", exception.reason);
    }
    
    ((void (*)(id, SEL, id))originalHandleMessage)(self, _cmd, message);
}

__attribute__((constructor))
static void initialize() {
    NSLog(@"[DouyinMysteryUser] dylib已加载");
    
    mysteryUserInfoCache = [NSMutableDictionary dictionary];
    
    @autoreleasepool {
        Class roomManagerClass = NSClassFromString(@"AwemeRoomMessageManager");
        if (roomManagerClass) {
            SEL handleMessageSel = NSSelectorFromString(@"handleMessage:");
            Method originalMethod = class_getInstanceMethod(roomManagerClass, handleMessageSel);
            
            if (originalMethod) {
                originalHandleMessage = method_getImplementation(originalMethod);
                method_setImplementation(originalMethod, (IMP)hooked_handleMessage);
                NSLog(@"[DouyinMysteryUser] Hook成功: AwemeRoomMessageManager handleMessage:");
            } else {
                NSLog(@"[DouyinMysteryUser] 未找到 handleMessage: 方法");
            }
        } else {
            NSLog(@"[DouyinMysteryUser] 未找到 AwemeRoomMessageManager 类");
        }
    }
}
